livekit-agents/livekit/agents/stt/fallback_adapter.py

```diff
@@ -0,0 +1,153 @@
+        self._connect_timeout = connect_timeout
+        self._keepalive_timeout = keepalive_timeout
+
+    async def recognize(
+        self, buffer: utils.AudioBuffer, *, language: str | None = None
+    ) -> SpeechEvent:
+        def _init_stt():
+            index = next((i for i, x in enumerate(self._providers) if x.active), None)
+            if isinstance(index, int):
+                return index
+            else:
+                raise Exception("all providers failed")
+
+        i = _init_stt()
+        while True:
+            try:
+                return await asyncio.wait_for(
+                    self._providers[i].stt.recognize(buffer, language=language),
+                    self._connect_timeout,
+                )
+            except (TimeoutError, asyncio.TimeoutError):
+                logger.warn(
+                    f"provider {self._providers[i].__class__.__module__} failed, attempting to switch"
+                )
+                self._providers[i].deactivate()
+                i = _init_stt()
+
+    def stream(self, *, language: str | None = None) -> SpeechStream:
+        if self.capabilities.streaming:
+            return FallbackSpeechStream(
+                providers=self._providers,
+                connect_timeout=self._connect_timeout,
+                keepalive_timeout=self._keepalive_timeout,
+                language=language,
+            )
+        raise NotImplementedError(
+            "streaming is not supported by this STT, please use a different STT or use a StreamAdapter"
+        )
+
+
+class FallbackSpeechStream(SpeechStream):
+    def __init__(
+        self,
+        *,
+        providers: list[STTProvider],
+        connect_timeout: float,
+        keepalive_timeout: float,
+        language: str | None,
+    ) -> None:
+        super().__init__()
+        self._connect_timeout = connect_timeout
+        self._keepalive_timeout = keepalive_timeout
+        self._providers = providers
+        self._language = language
+        self._timeout: float
+        self._init_stt()
+
+    @utils.log_exceptions(logger=logger)
+    def _init_stt(self) -> None:
+        index = next((i for i, x in enumerate(self._providers) if x.active), None)
+        if index is not None:
+            self._index = index
+            self._stream = self._providers[self._index].stt.stream(
+                language=self._language
+            )
+            self._timeout = self._connect_timeout
+        else:
+            raise Exception("all providers failed")
+
+    @utils.log_exceptions(logger=logger)
+    async def _main_task(self) -> None:
+        async def _pass_input():
+            async for input in self._input_ch:
+                self._stream.push_frame(input)
+
+        async def _pass_output():
+            while True:
+                try:
+                    item = await asyncio.wait_for(
+                        self._stream.__anext__(), self._timeout
```

Comment:
Not sure if timeout makes sense for a streamed STT. The user could just "not speak" and we would reach this timeout

Reply:
this is kinda the same issue as the other StreamAdapter; we don't have a keepalive function so we can't discern between standby and timeout. i want to figure out a good solution for this on the TTS side first then port it here

Reply:
we still have a way of detecting if the connection is alive right? maybe it's better to rely on connection-level timeout first, to eliminate false positives

Reply:
`StreamAdapter`s don't use a websocket, because they just map normal requests to fit the `SynthesizeStream` object, so we can't check for that either unfortunately